sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java

```diff
@@ -347,12 +347,12 @@ public boolean equals(Object o) {
     }
     Row other = (Row) o;
     return Objects.equals(getSchema(), other.getSchema())
-        && Objects.equals(getValues(), other.getValues());
+        && Objects.deepEquals(getValues().toArray(), other.getValues().toArray());
```

Comment:
This is actually not enough. If you have a schema with `f: LIST<BYTES>` it will not have correct equality. We need our own schema-driven deep equality.

Reply:
Yes. This is a valid concern. If there is no other option, we need our deep equality check for our schema.

https://issues.apache.org/jira/browse/BEAM-5868

Reply:
The problem comes from `Objects.deepEquals` only have a deep equal implementation for primitive types and array. So Map and List will at least fail to check the correct equality.